### PR TITLE
Add second message for LiveDrop errors

### LINKED ISSUE #72907: Unhelpful "constants cannot evaluate destructors" with loops
This code:
```rust
#![feature(const_if_match)]
#![feature(const_loop)]

const _: Option<Vec<i32>> = {
    let mut never_returned = Some(Vec::new());
    let mut always_returned = None;

    let mut i = 0;
    loop {
        always_returned = never_returned;
        never_returned = None;

        i += 1;
        if i == 10 {
            break always_returned;
        }
    
    }
};
```
yields the following error:
```
error[E0493]: destructors cannot be evaluated at compile-time
 --> src/lib.rs:7:9
  |
7 |     let mut always_returned = None;
  |         ^^^^^^^^^^^^^^^^^^^ constants cannot evaluate destructors

error: aborting due to previous error
```
That is not very helpful as there is nothing with a destructor on that line. The actual line where something could get dropped is line 11, and that is also the line where the drop-requiring thing is assigned into `always_returned`.

Cc @rust-lang/wg-const-eval

### PR DESCRIPTION
**Pull request by pvdrz**
This is an attempt to fix https://github.com/rust-lang/rust/issues/72907 by adding a second message to the `LiveDrop` diagnostics. Changing from this
```
error[E0493]: destructors cannot be evaluated at compile-time
 --> src/lib.rs:7:9
  |
7 |     let mut always_returned = None;
  |         ^^^^^^^^^^^^^^^^^^^ constants cannot evaluate destructors

error: aborting due to previous error
```
to this
```
error[E0493]: destructors cannot be evaluated at compile-time
  --> foo.rs:6:9
   |
6  |     let mut always_returned = None;
   |         ^^^^^^^^^^^^^^^^^^^ constants cannot evaluate destructors
...
10 |         always_returned = never_returned;
   |         --------------- value is dropped here

error: aborting due to previous error
```
r? @RalfJung @ecstatic-morse

### CODE REVIEW

File: src/librustc_mir/transform/check_consts/validation.rs [hunk 1]
```diff
@@ -588,7 +588,13 @@ impl Visitor<'tcx> for Validator<'mir, 'tcx> {
                 };
 
                 if needs_drop {
-                    self.check_op_spanned(ops::LiveDrop, err_span);
+                    let term_span = terminator.source_info.span;
+                    let op = if term_span.hi() - term_span.lo() > rustc_span::BytePos(1) {
```
Comment (oli-obk) [inline src/librustc_mir/transform/check_consts/validation.rs:592]: I think what you want is `!term_span.is_dummy()` at which point you should be able to not do this branch at all, as the diagnostics system can handle dummy spans I think

Comment (oli-obk) [reply src/librustc_mir/transform/check_consts/validation.rs:592]: The ui test diffs seem good to me and you can probably just bless them.

Can you also add a test showing your example in the PR post?

Comment (christianpoveda) [reply src/librustc_mir/transform/check_consts/validation.rs:592]: I was just following @ecstatic-morse suggestion: https://github.com/rust-lang/rust/issues/72907#issuecomment-637821922

Also without it there are a lot of overlapping messages, let me get you an example

File: src/test/ui/consts/miri_unleashed/feature-gate-unleash_the_miri_inside_of_you.stderr [hunk 2]
```diff
@@ -2,7 +2,9 @@ error[E0493]: destructors cannot be evaluated at compile-time
   --> $DIR/feature-gate-unleash_the_miri_inside_of_you.rs:11:20
    |
 LL |     const F: u32 = (U::X, 42).1;
-   |                    ^^^^^^^^^^ constants cannot evaluate destructors
+   |                    ^^^^^^^^^^ - value is dropped here
```
Comment (oli-obk) [inline src/test/ui/consts/miri_unleashed/feature-gate-unleash_the_miri_inside_of_you.stderr:5]: This is a slightly odd span. It points to the field index. I mean it's kind of correct, the field access is where it gets dropped... Probably fine to leave it as it is

**Pull request merged**